packages/builders/src/messages/Assertions.ts

```diff
@@ -27,40 +27,49 @@ export const messageReferencePredicate = z.object({
 	type: z.nativeEnum(MessageReferenceType).optional(),
 });
 
-export const messagePredicate = z
-	.object({
+const baseMessagePredicate = z.object({
+	nonce: z.union([z.string().max(25), z.number()]).optional(),
+	tts: z.boolean().optional(),
+	allowed_mentions: allowedMentionPredicate.optional(),
+	message_reference: messageReferencePredicate.optional(),
+	attachments: attachmentPredicate.array().max(10).optional(),
+	enforce_nonce: z.boolean().optional(),
+});
+
+const basicActionRowPredicate = z.object({
+	type: z.literal(ComponentType.ActionRow),
+	components: z
+		.object({
+			type: z.union([
+				z.literal(ComponentType.Button),
+				z.literal(ComponentType.ChannelSelect),
+				z.literal(ComponentType.MentionableSelect),
+				z.literal(ComponentType.RoleSelect),
+				z.literal(ComponentType.StringSelect),
+				z.literal(ComponentType.UserSelect),
+			]),
+		})
+		.array(),
+});
+
+const messageNoComponentsV2Predicate = baseMessagePredicate
+	.extend({
 		content: z.string().optional(),
-		nonce: z.union([z.string().max(25), z.number()]).optional(),
-		tts: z.boolean().optional(),
 		embeds: embedPredicate.array().max(10).optional(),
-		allowed_mentions: allowedMentionPredicate.optional(),
-		message_reference: messageReferencePredicate.optional(),
-		// Partial validation here to ensure the components are valid,
-		// rest of the validation is done in the action row predicate
-		components: z
-			.object({
-				type: z.literal(ComponentType.ActionRow),
-				components: z
-					.object({
-						type: z.union([
-							z.literal(ComponentType.Button),
-							z.literal(ComponentType.ChannelSelect),
-							z.literal(ComponentType.MentionableSelect),
-							z.literal(ComponentType.RoleSelect),
-							z.literal(ComponentType.StringSelect),
-							z.literal(ComponentType.UserSelect),
-						]),
-					})
-					.array(),
-			})
-			.array()
-			.max(5)
-			.optional(),
 		sticker_ids: z.array(z.string()).min(0).max(3).optional(),
-		attachments: attachmentPredicate.array().max(10).optional(),
-		flags: z.number().optional(),
-		enforce_nonce: z.boolean().optional(),
 		poll: pollPredicate.optional(),
+		components: basicActionRowPredicate.array().max(5).optional(),
+		flags: z
+			.number()
+			.optional()
+			.refine((flags) => {
+				// If we have flags, ensure we don't have the ComponentsV2 flag
+				if (flags) {
+					return (flags & MessageFlags.IsComponentsV2) !== MessageFlags.IsComponentsV2;
```

Comment:
This feels more idiomatic as it is not an inverse check:

```suggestion
					return (flags & MessageFlags.IsComponentsV2) === 0
```